`spy_messages_to_l1` - Docs (#2284)
foundry-rs/starknet-foundry: docs/src/testing/testing-messages-to-l1.md

# Testing messages to l1

There exists analogous functionality to [spying events](./testing-events.md) for messages sent to L1.

[Check the appendix for an exact API, structures and traits reference](../appendix/cheatcodes/spy_messages_to_l1.md)

Asserting messages to L1 is much simpler, since they are not wrapped with any structures in cairo code (they are plain felt array and an address). 
In `snforge` they are expressed with a structure:

```rust
/// Raw message to l1 format (as seen via the RPC-API), can be used for asserting the sent messages.
struct MessageToL1 {
    /// An ethereum address where the message is destined to go
    to_address: EthAddress,
    /// Actual payload which will be delivered to L1 contract
    payload: Array<felt252>
}
```

Similarly, you can use starknet foundry's `spy_messages_to_l1()` to initiate a spy:

```rust
use snforge_std::{spy_messages_to_l1};
// ...
let mut spy = spy_messages_to_l1();
```

> Then, you can execute some code, and make the assertions:

Reviewer: ```suggestion
With the spy ready to use, you can execute some code, and make the assertions:
```

1. Either with the spy directly (via `AssertionsTrait` API) like so:

```rust
use snforge_std::{spy_messages_to_l1, MessageToL1SpyAssertionsTrait, MessageToL1};

spy.assert_sent(
    @array![
        (
            contract_address, // Message sender
            MessageToL1 {     // Message content (receiver and payload)
                to_address: 0x123.try_into().unwrap(), 
                payload: array![123, 321, 420]
            }
        )
    ]
);
```

2. Or use the messages directly via `get_messages()` API of the `MessageToL1Spy`:

```rust

> use snforge_std::{
>     spy_messages_to_l1, MessageToL1, 
>     MessageToL1SpyAssertionsTrait, 
>     MessageToL1FilterTrait, 
>     MessageToL1SpyTrait
> };
> 
> let messages = spy.get_messages();
> // Use filtering optionally on MessagesToL1 instance
> let messages_from_specific_address = messages.sent_by(sender_address);
> let messages_to_specific_address   = messages_from_specific_address.sent_to(receiver_eth_address);
> 
> // Get the messages from the MessagesToL1 structure 
> let (from, message) = messages_to_specific_address.messages.at(0);
> 
> // Assert the sender
> assert!(from == sender_address, "Sent from wrong address");
> // Assert the MessageToL1 fields
> assert!(message.to_address == receiver_eth_address, "Wrong eth address of the receiver");
> assert!(message.payload.len() == 3, "There should be 3 items in the data");
> assert!(*message.payload.at(1) == 421, "Expected 421 in payload");

Reviewer: ```suggestion
use snforge_std::{
    spy_messages_to_l1, MessageToL1, 
    MessageToL1SpyAssertionsTrait, 
    MessageToL1FilterTrait, 
    MessageToL1SpyTrait
};

// ...

#[test]
fn test_spying_l1_messages() {
    let mut spy = spy_messages_to_l1();
    
    let messages = spy.get_messages();
    
    // Use filtering optionally on MessagesToL1 instance
    let messages_from_specific_address = messages.sent_by(sender_address);
    let messages_to_specific_address   = messages_from_specific_address.sent_to(receiver_eth_address);

    // Get the messages from the MessagesToL1 structure 
    let (from, message) = messages_to_specific_address.messages.at(0);

    // Assert the sender
    assert!(from == sender_address, "Sent from wrong address");
    // Assert the MessageToL1 fields
    assert!(message.to_address == receiver_eth_address, "Wrong eth address of the receiver");
    assert!(message.payload.len() == 3, "There should be 3 items in the data");
    assert!(*message.payload.at(1) == 421, "Expected 421 in payload");
```

```

> 

Reviewer: ```suggestion
```
